### PR TITLE
refactor(connlib): drop all connections when roaming

### LINKED ISSUE #5080: connlib: implement reconnect as "drop all connections and wait for new packets to trigger new ones"
Currently, connlib stores a fair bit of connection-specific state. Most of it is for DNS resources because the resolved IPs are bound to the gateway that we connected to. With https://github.com/firezone/firezone/issues/4994, this changes and we are implementing a persistent DNS resolver that is independent of the gateways we are connected to.

The only other pieces of state that we still store per connection on the client are:

- "Allowed IPs", a (legacy) concept inherited from how wireguard is implemented in the kernel and boringtun. Because we have ephemeral tunnels, this doesn't actually do anything useful for us and is bound to be removed in https://github.com/firezone/firezone/pull/5077
- Tracking of mangled DNS requests to resolvers that are also resources. This will be moved out of connection-specific state in https://github.com/firezone/firezone/issues/5079.

With no connection-specific state left for clients, roaming can be implemented in a much simpler way: Drop all connections and simply establish new ones as packets are arriving. Whilst a client is roaming, packets will always be lost simply because there isn't any connectivity yet, thus, we might as well use those packets to generate new connection intents and connect to the correct gateway via the best possible path.

Currently, snownet tries to be clever in preserving this connection state and migrate the connection to a new endpoint. This however is very tricky to get right and thus error-prone. For example, if we are on a relayed connection (because of restrictive NAT or otherwise limited connectivity) and we roam to a new one where hole-punching is possible, we actually want connlib to now use that better network path. But how do we know that? The ICE spec has the concept of ICE restart for this which essentially clears the local connection state and re-triggers the candidate selection process. Ultimately, for us there isn't a difference in perform ICE restart or throwing away the entire connection and making a new one.

### PR DESCRIPTION
**Pull request by thomaseizinger**
Currently, `snownet` tries to be very clever in how it roams connections. This is/was necessary because we associated DNS-specific state with a connection. More specifically, the assigned proxy IPs for a DNS resource are stored as part of a connection with the gateway.

As a result, DNS resources would always break if the underlying connection in `snownet` failed. This is quite error prone and means, `snownet` must be very careful to never-ever fail a connection erroneously. With #5049, we no longer store any important state with a connection and thus, can implement roaming in much simpler way: Drop all connections and let the incoming packets create new ones. This is much more robust as we don't have to "patch" existing state in `snownet` as part of roaming.

We test this new functionality by adding a `RoamClient` transition to `tunnel_test`. This ensures roaming works in a lot of scenarios, including relayed and non-relayed situations as well as roaming between either of them. As a result, we can delete several of the more specific test cases of `snownet`.

Depends-On: #5049.
Replaces: #5060.
Resolves: #5080.

### CODE REVIEW

**Comment by thomaseizinger** [comment]
This already works for CIDR resources but for DNS resources, we still need to wait on #5049.

File: scripts/tests/direct-download-roaming-network.sh
Comment (thomaseizinger) [inline scripts/tests/direct-download-roaming-network.sh:1]: Had to slightly adopt this download test: `curl` appears to not send any packets as part of downloading and thus once interrupted due to roaming, the connection does not get re-established. I added `--retry 1` and `--continue-at -` to retry the download on failure. Previously, the file was just appended via stdout which failed the hash-check after the retry. Now we use `curl`'s `--output` which means we also need to calculate the hash in the container.

File: rust/connlib/tunnel/src/client.rs [hunk 1]
```diff
@@ -818,6 +818,15 @@ impl ClientState {
                             resources: self.resources(),
                         });
                 }
+                snownet::Event::ConnectionsCleared(ids) => {
+                    for id in ids {
+                        self.cleanup_connected_gateway(&id);
+                    }
```
Comment (conectado) [inline rust/connlib/tunnel/src/client.rs:824]: a `debug_assert!` here that all peers are empty would be nice I think

File: scripts/tests/direct-download-roaming-network.sh
Comment (conectado) [reply scripts/tests/direct-download-roaming-network.sh:1]: We should have a user-facing documentation of this behavior.

We want users to know that downloads won't resume after roaming automatically, I think this is okay and expected for TCP connections, but it's still good to document.

**Review by conectado — APPROVED** [review]
Nice!

**Pull request merged**